README
NAME
    ThreatNet::Filter - Interface for ThreatNet event filters

DESCRIPTION
    ThreatNet data sources can potentially generate a hell of a lot of
    events, and it's important to be able to filter these down to just the
    events that matter.

    Many of the filters are stateful. For example, the standard
    ThreatNet::Filter::ThreatCache module provides cache objects that filter
    out any threats that have already been seen in the previous hour. (or
    whatever the state period is).

METHODS
    The filter API is quite simple, with only a few methods.

  new ...
    Since some categories of filter do not strictly need to be in the form
    of an object, a default "new" constructor is provided which just creates
    an empty object.

    Returns a new "ThreatNet::Filter" object, or some sub-classes may return
    "undef" on error.

  keep $Message
    The "keep" method takes a ThreatNet::Message object and examines it to
    determine if the message should be kept, or filtered out.

    In the default implementation of the filter, all messages are kept.

    Returns true if the message should be kept, or false if the message
    should be discarded.

SUPPORT
    All bugs should be filed via the bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ThreatNet-Filter>

    For other issues, or commercial enhancement and support, contact the
    author

AUTHORS
    Adam Kennedy (Maintainer), <http://ali.as/>, cpan@ali.as

SEE ALSO
    <http://ali.as/threatnet/>, ThreatNet::Message

COPYRIGHT
    Copyright (c) 2004 - 2005 Adam Kennedy. All rights reserved. This
    program is free software; you can redistribute it and/or modify it under
    the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension ThreatNet::Filter

0.04 Wed Jun 15 2005
	- Fixing a small stats bug

0.03 Sun Jun 12 2005
	- Added statistics collection to ThreatNet::Filter::ThreatCache

0.02 Sat May  7 2005
	- Added ThreatNet::Filter::Network

0.01 Sun May  1 2005
	- Created the initial dist with the abstrct, null,
	  chain and Threat Cache filters.


lib/ThreatNet/Filter.pm
package ThreatNet::Filter;

=pod

=head1 NAME

ThreatNet::Filter - Interface for ThreatNet event filters

=head1 DESCRIPTION

ThreatNet data sources can potentially generate a hell of a lot of
events, and it's important to be able to filter these down to just
the events that matter.

Many of the filters are stateful. For example, the standard
L<ThreatNet::Filter::ThreatCache> module provides cache objects
that filter out any threats that have already been seen in the
previous hour. (or whatever the state period is).

=head1 METHODS

The filter API is quite simple, with only a few methods.

=cut

use strict;
use Params::Util '_INSTANCE';

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.04';
}





#####################################################################
# Constructor

=pod

=head2 new ...

Since some categories of filter do not strictly need to be in the
form of an object, a default C<new> constructor is provided which just
creates an empty object.

Returns a new C<ThreatNet::Filter> object, or some sub-classes may
return C<undef> on error.

=cut

sub new {
	bless {}, shift;
}

=pod

=head2 keep $Message

The C<keep> method takes a L<ThreatNet::Message> object and examines
it to determine if the message should be kept, or filtered out.

In the default implementation of the filter, all messages are kept.

Returns true if the message should be kept, or false if the message
should be discarded.

=cut

sub keep { _INSTANCE($_[1], 'ThreatNet::Message') ? 1 : undef }

1;

=pod

=head1 SUPPORT

All bugs should be filed via the bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ThreatNet-Filter>

For other issues, or commercial enhancement and support, contact the author

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 SEE ALSO

L<http://ali.as/threatnet/>, L<ThreatNet::Message>

=head1 COPYRIGHT

Copyright (c) 2004 - 2005 Adam Kennedy. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


lib/ThreatNet/Filter/Null.pm
package ThreatNet::Filter::Null;

=pod

=head1 NAME

ThreatNet::Filter::Null - ThreatNet Filter to discard all messages

=head1 DESCRIPTION

The default C<ThreatNet::Filter> object returns true for all messages
passed to its C<keep> method, and can thus be used as a logical "true"
filter object if needed.

C<ThreatNet::Filter::Null> is a utility class which provides the logical
opposite of this. Calls to the C<keep> method for Null filter objects
B<always> return false.

Its the bit bucket of the ThreatNet Filter world.

=head1 METHODS

Methods are as for the parent L<ThreatNet::Filter> class.

=cut

use strict;
use Params::Util '_INSTANCE';
use base 'ThreatNet::Filter';

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.04';
}

sub keep {
	_INSTANCE($_[1], 'ThreatNet::Message') ? '' : undef;
}

1;

=pod

=head1 SUPPORT

All bugs should be filed via the bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ThreatNet-Filter>

For other issues, or commercial enhancement and support, contact the author

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 SEE ALSO

L<http://ali.as/threatnet/>, L<ThreatNet::Filter>

=head1 COPYRIGHT

Copyright (c) 2005 Adam Kennedy. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


lib/ThreatNet/Filter/Chain.pm
package ThreatNet::Filter::Chain;

=pod

=head1 NAME

ThreatNet::Filter::Chain - Create a chain of ThreatNet filters

=head1 SYNOPSIS

    # After stripping junk ips, pass events through a Threat Cache,
    # and then keep only the events that have occured in an IP range
    # that we are responsible for.
    my $Chain = ThreatNet::Filter::Chain->new(
    	ThreatNet::Filter::Junk->new,
    	ThreatNet::Filter::ThreatCache->new,
    	ThreatNet::Filter::Network->new( '202.123.123.0/24' ),
    	);
    
    sub process_message {
    	my $Message = shift;
    	
    	unless ( $Chain->keep($Message) ) {
    		return;
    	}
    	
    	print "Threat spotted in our network at " . $Message->ip . "\n";
    }

=head1 DESCRIPTION

C<ThreatNet::Filter::Chain> lets you create filters that represent an
entire chain of filters. L<ThreatNet::Message> objects are checked against
each individual filter in the same order.

A message must pass the C<keep> method of each filter to move down the
chain. If a message is rejected at a point in chain, filters further down
the chain will not see them. This is mainly of importance to stateful
filters such as L<ThreatNet::Filter::ThreatCache>.

It is assumed you don't actual care B<which> filter rejects a message,
and as such there is no way to tell this :)

=head1 METHODS

The methods are the same as for the parent L<ThreatNet::Filter> class, but
with a change to the C<new> constructor.

=cut

use strict;
use Params::Util '_SET', '_INSTANCE';
use base 'ThreatNet::Filter';

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.04';
}

=pod

=head2 new $Filter [, $Filter, ... ]

The C<new> constructor takes a set of (1 or more) L<ThreatNet::Filter>
objects, and creates a Filter object that acts as a I<"short-cutting
logical AND"> combination of them.

Returns a new C<ThreatNet::Filter::Chain> object, or C<undef> if not
provided with the correct params.

=cut

sub new {
	my $class   = ref $_[0] ? ref shift : shift;
	my $filters = _SET([ @_ ], 'ThreatNet::Filter') or return undef;

	my $self  = $class->SUPER::new;
	$self->{filters} = $filters;

	$self;
}

=pod

=head2 keep $Message

The C<keep> method takes a single L<ThreatNet::Message> object and checks
it against each of the child filters in turn, short-cutting if any of
them does not want to keep the message.

One small note - The C<keep> method returns B<exactly> the same false value
it recieves from the child filter, whether that is normal false or C<undef>.

Returns true if you should keep the message, or false to discard.

=cut

sub keep {
	my $self    = shift;
	my $Message = _INSTANCE(shift, 'ThreatNet::Message') or return undef;

	foreach my $Filter ( @{$self->{filters}} ) {
		my $rv = $Filter->keep($Message);
		return $rv unless $rv; # false or undef
	}

	1;
}

1;

=pod

=head1 SUPPORT

All bugs should be filed via the bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ThreatNet-Filter>

For other issues, or commercial enhancement and support, contact the author

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 SEE ALSO

L<http://ali.as/threatnet/>, L<ThreatNet::Filter>

=head1 COPYRIGHT

Copyright (c) 2005 Adam Kennedy. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
	

lib/ThreatNet/Filter/ThreatCache.pm
package ThreatNet::Filter::ThreatCache;

=pod

=head1 NAME

ThreatNet::Filter::ThreatCache - A Threat Cache implementated as a filter

=head1 DESCRIPTION

A C<ThreatNet::Filter::ThreatCache> is a basic filter-based implementation
of a I<Threat Cache>, as defined in the ThreatNet concept paper.
(L<http://ali.as/threatnet/>)

The consistent use of Threat Caches by all nodes is the key to keeping
message quantities to a minimum, and allows the entire network to safely
run without any canonical state.

As each message is provided to the filter, it stores an IPs that it sees,
filtering out any ips that have already been seen in the last hour (or
custom period if provided).

=head1 METHODS

=cut

use strict;
use Params::Util '_INSTANCE';
use base 'ThreatNet::Filter';

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.04';
}





#####################################################################
# ThreatNet::Filter Interface

=pod

=head2 new [ $param => $value, ... ]

The C<new> method creates a new Threat Cache object.

It takes a single optional parameter C<timeout> which should the
positive integer number of seconds the channel dictates as the mimumum
time before an event can be rementioned. The default value is 3600
(1 hour).

Returns a new C<ThreatNet::Filter::ThreatCache> object.

=cut

sub new {
	my $class = ref $_[0] ? ref shift : shift;
	my %args  = @_;

	# Add the cache attributes to the base filter object
	my $self = $class->SUPER::new;
	$self->{cache_ip}   = {};
	$self->{cache_time} = [];

	# When we will be syncronised?
	# Add a couple of seconds to allow for small initialization delays.
	my $t = time();
	$self->{timeout} = defined $args{timeout} ? $args{timeout} : 3600;
	$self->{sync_at} = $self->{timeout} + $t + 2;

	# Collect some very basic statistic
	$self->{stats} = { time_start => $t, => seen => 0, kept => 0 };

	$self;
}

=pod

=head2 keep $Message

As in the parent L<ThreatNet::Filter> class, the C<keep> method takes as
argument a single C<ThreatNet::Message> object.

It returns true if the message can be kept, or false if the message should
be dropped.

=cut

sub keep {
	my $self    = shift;
	my $Message = _INSTANCE(shift, 'ThreatNet::Message') or return undef;
	$Message->can('ip') or return undef; # WTF is it if it can't?

	# Flush old events out of the cache
	while ( my $event = $self->{cache_time}->[0] ) {
		if ( time() > $event->{time} + $self->{timeout} ) {
			# Block has expired
			shift @{$self->{cache_time}};
			delete $self->{cache_ip}->{$event->{ip}};
		} else {
			# Because the events are added (we assume) in creation
			# order, as soon as we encounter one that has not expired,
			# we can assume the rest have not expired.
			last;
		}
	}

	# We only want the time and IP
	### Convert this to ->event_time later?
	my $created = $Message->created or return undef;
	my $ip      = $Message->ip      or return undef;

	# At this point, we've officially "seen" the message
	$self->{stats}->{seen}++;

	# If the ip is in the cache, don't keep it
	return '' if $self->{cache_ip}->{$ip};

	# Add to the cache and signal to keep.
	### Stricly speaking, just dropping it onto the end of the
	### queue is not good enough. Convert this to something that
	### ensures correct order once there is a chance that the events
	### may come in out of order, such as if we change from object
	### creation time to event time.
	$self->{cache_ip}->{$ip} = $created;
	push @{$self->{cache_time}}, { time => $created, ip => $ip };
	$self->{stats}->{kept}++;

	1;
}

=pod

=head2 synced

The C<synced> method checks to see if the Threat Cache has synchronised
with the channel.

Returns true if the current time is past the sync time, or false if not.

=cut

sub synced {
	my $self = shift;
	!! (time() > $self->{sync_at});
}

=pod

=head2 stats

The C<stats> method returns a hash with a variety of statistics from
the Threat Cache.

Returns the stats as a C<HASH> reference.

=cut

sub stats {
	my $self  = shift;
	my %stats = ();

	# Generate the general statistics
	$stats{size}    = scalar @{$self->{cache_time}};
	$stats{seen}    = $self->{stats}->{seen};
	$stats{kept}    = $self->{stats}->{kept};
	$stats{discard} = $stats{seen} - $stats{kept};
	$stats{expired} = $stats{kept} - $stats{size};
	
	# Add the time statistics
	$stats{time_start}   = $self->{stats}->{time_start};
	$stats{time_current} = time();
	$stats{time_running} = $stats{time_current} - $stats{time_start};

	# Percentages
	$stats{percent_kept}    = $self->_perc($stats{kept}, $stats{seen});
	$stats{percent_discard} = $self->_perc($stats{discard}, $stats{seen});

	# Rates
	$stats{rate_seen} = $self->_rate($stats{seen}, $stats{time_running});
	$stats{rate_kept} = $self->_rate($stats{kept}, $stats{time_running});

	\%stats;
}

sub _perc {
	my (undef, $items, $total) = @_;
	my $perc = $total ? ($items / $total) : 0;
	$perc = $perc * 100;
	sprintf("%0.1f", $perc) . '%';
}

sub _rate {
	my (undef, $items, $interval) = @_;
	my $rate = $interval ? ($items / $interval) : 0;
	sprintf("%0.1f", $rate);
}

1;

=pod

=head1 SUPPORT

All bugs should be filed via the bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ThreatNet-Filter>

For other issues, or commercial enhancement and support, contact the author

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 SEE ALSO

L<http://ali.as/devel/threatnetwork.html>, L<ThreatNet::Filter>

=head1 COPYRIGHT

Copyright (c) 2005 Adam Kennedy. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


lib/ThreatNet/Filter/Network.pm
package ThreatNet::Filter::Network;

=pod

=head1 NAME

ThreatNet::Filter::Network - Filter events within a set of IP ranges

=head1 SYNOPSIS

  # Filter for IPs in our broadband customers' range
  my $Broadband = ThreatNet::Filter::Network->new( keep => '123.123.0.0/16' );
  
  # Create a filter for "local" and other things we want to discard
  # (including our own personal broadband IP in the above network)
  my $NotLocal = ThreatNet::Filter::Network->new( discard => qw{
      123.123.123.123
      LOCAL
      111.245.76.248/29
      222.234.52.192/29
      } );
  
  sub boot_zombies {
  	my $Msg = shift;
  	if ( $Broadband->keep($Msg) and $NotLocal->keep($Msg) ) {
  		my $account = $RadiusServer->ip_to_account($Msg->ip);
  		$account->disable();
  		$account->disconnect();
  		$account->add_support_note("You are infected with a virus");
  	}
  }

=head1 DESCRIPTION

C<ThreatNet::Filter::Network> is a filter class for creating network
filters.

That is, for filtering event streams to just those events that did
(or did not) occur within a particular network.

The objects only check in two modes.

The C<keep> keyword as first argument indicates events should be kept if
they match B<any> of the networks.

The C<discard> keyword as first argument indicates events should be kept
only if they do B<not> match any of networks.

For more complex network masks, see the L<ThreatNet::Filter::Chain> class
for chaining groups of C<keep> and C<discard> filters together.

=head2 Specifying the Networks

The actual matching is done using the L<Net::IP::Match::XS> module. Any
values that can be used by it can also be used with it can thus also be
used with C<ThreatNet::Filter::Network>.

=head2 Keyword Expansion

In addition to the normal IP specification above,
C<ThreatNet::Filter::Network> also supports keyword expansion for a
number of standard sets of network masks.

When specified by name, they will be expanded into a list of IP ranges.

Thus you can do something like the following.

  my $Remove = ThreatNet::Filter::Network->new(
      discard => 'RFC1918', '123.123.123.0/24'
      );

This will filter out the three standard "local" IP blocks specified by
RFC1918, plus the addition range 123.123.123.0 - 123.123.123.255.

All keywords are case-insensitive.

=head3 RFC1918

The C<RFC1918> keyword is expanded to the three network blocks reserved
for local intranets. This specifically does NOT include the localhost
address space.

=head3 RFC3330

The C<RFC3330> keyword is expanded to a larger set of network blocks
restricted for various purposes as identifier in RFC3330. This includes
those from C<RFC1918>, the localhost block, and several additional
blocks reserved for benchmarking, IP 6to4 identifiers and various other
blocks that should not appear in threat messages.

Where correctness is a factor, such as posting to a non-C<tolerant>
channel, this filter should be applied before issuing messages, as they
are highly likely to be fraudulent or technically nonsensical.

=head3 LOCAL

The C<LOCAL> keyword is expanded to represent the most common
interpretation of a "local" address, which is the RFC1918 addresses,
plus the C<127.0.0.0/8> localhost block.

=head2 Message Compatibility

Please note that because the module on which this filter is based only
supports IPv4 ranges, this filter class is B<only> capable of processing
L<ThreatNet::Message::IPv4> (or subclass) objects.

Any other message types passed to C<keep> will be returns C<undef>, and
thus will act as a null filter in most configurations.

=head1 METHODS

=cut

use strict;
use Params::Util '_INSTANCE',
                 '_IDENTIFIER';
use base 'ThreatNet::Filter';
use Net::IP::Match::XS ();
use ThreatNet::Message::IPv4 ();

use vars qw{$VERSION %KEYWORD};
BEGIN {
	$VERSION = '0.04';

	# Expandable network keywords
	%KEYWORD = ();
	$KEYWORD{RFC1918} = [qw{
		10.0.0.0/8
		172.16.0.0/12
		192.168.0.0/16
		}];
	$KEYWORD{RFC3330} = [qw{
		0.0.0.0/8
		10.0.0.0/8
		127.0.0.0/8
		169.254.0.0/16
		172.16.0.0/12
		192.0.2.0/24
		192.168.0.0/16
		192.88.99.0/24
		198.18.0.0/15
		224.0.0.0/4
		240.0.0.0/4
		}];
	$KEYWORD{LOCAL} = [qw{
		127.0.0.0/8
		10.0.0.0/8
		172.16.0.0/12
		192.168.0.0/16
		}];			
}





####################################################################3
# Constructor and Accessors

=pod

=head2 new ('keep' | 'discard'), $network, ...

The C<new> constructor takes a param of either C<keep> or C<discard>,
followed by a list of one or more values which are either an expandable
keyword or an ip ranges compatible with L<Net::IP::Match::XS>.

A ThreatNet filter is created which limits a message stream to events
either inside or outside of the resulting network.

Returns a new C<ThreatNet::Filter::Network> object, or C<undef> if given
invalid params.

=cut

sub new {
	my $class = ref $_[0] ? ref shift : shift;

	# Check we are either a 'keep' or 'discard' filter
	my $type  = lc _IDENTIFIER(shift)     or return undef;
	$type eq 'keep' or $type eq 'discard' or return undef;

	# Get the list of values, filter and keyword-expand
	my @network = map { $KEYWORD{$_} ? @{$KEYWORD{$_}} : $_ }
		map { uc $_ } grep { defined $_ } @_;

	# Initial version doesn't check the IP ranges.
	# It just makes sure we have at least one.
	@network or return undef;

	# Create the object
	my $self = bless {
		keep    => $type eq 'keep',
		network => [ @network ],
		}, $class;

	$self;
}

=pod

=head2 type

The C<type> accessor returns the type of the network filter.

Returns either C<'keep'> or C<'discard'>.

=cut

sub type {
	$_[0]->{keep} ? 'keep' : 'discard';
}

=pod

=head2 network

The C<network> accessor returns the list of ip ranges as provided to
the constructor.

=cut

sub network {
	@{$_[0]->{network}};
}





#####################################################################
# ThreatNet::Filter Methods

=pod

=head2 keep $Message

The C<keep> method takes a C<ThreatNet::Message::IPv4> message as per the
L<ThreatNet::Filter> specification, and checks it against the network
specification and C<keep>|C<discard> type.

Returns true if the message should be kept, false if not, or C<undef> on
error.

=cut

sub keep {
	my $self    = shift;
	my $Message = _INSTANCE(shift, 'ThreatNet::Message::IPv4') or return undef;
	my $rv = Net::IP::Match::XS::match_ip( $Message->ip, @{$self->{network}} );
	defined $rv ? $self->{keep} ? !! $rv : ! $rv : undef;
}

1;

=pod

=head1 SUPPORT

All bugs should be filed via the bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ThreatNet-Filter>

For other issues, or commercial enhancement and support, contact the author

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 SEE ALSO

L<http://ali.as/devel/threatnetwork.html>, L<ThreatNet::Filter>,
L<ThreatNet::Message::IPv4>.

=head1 COPYRIGHT

Copyright (c) 2005 Adam Kennedy. All rights reserved.
This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
